README.md
# ripgrep.el

[![License GPL 2][badge-license]][LICENSE]
[![Coverage Status](https://coveralls.io/repos/nlamirault/ripgrep.el/badge.png?branch=master)](https://coveralls.io/r/nlamirault/ripgrep.el?branch=master)

Master :
* [![MELPA Stable](https://stable.melpa.org/packages/ripgrep-badge.svg)](https://stable.melpa.org/#/ripgrep)
* [![Circle CI](https://circleci.com/gh/nlamirault/ripgrep.el/tree/master.svg?style=svg)](https://circleci.com/gh/nlamirault/ripgrep.el/tree/master)

Develop:
* [![Melpa Status](https://melpa.org/packages/ripgrep-badge.svg)](https://melpa.org/#/ripgrep)
* [![Circle CI](https://circleci.com/gh/nlamirault/ripgrep.el/tree/develop.svg?style=svg)](https://circleci.com/gh/nlamirault/ripgrep.el/tree/develop)

``ripgrep.el`` allows you to search using [ripgrep][] from inside Emacs.

## Installation

The recommended way to install ``ripgrep`` is via [MELPA][]:

    M-x package-install ripgrep

or [Cask][]:

	(depends-on "ripgrep")


There is also [projectile-ripgrep](https://melpa.org/#/projectile-ripgrep) if you want to use it using [projectile](https://github.com/bbatsov/projectile)


## Usage

<kbd>M-x ripgrep-regexp</kbd> or <kbd>M-x projectile-ripgrep</kbd>



## Development

### Cask

``ripgrep.el`` use [Cask][] for dependencies management. Install it and
retrieve dependencies :

    $ curl -fsSkL https://raw.github.com/cask/cask/master/go | python
    $ export PATH="$HOME/.cask/bin:$PATH"
    $ cask


### Testing

* Launch unit tests from shell

    $ make clean test

* Using [overseer][] :

Keybinding           | Description
---------------------|------------------------------------------------------------
<kbd>C-c , t</kbd>   | launch unit tests from buffer
<kbd>C-c , b</kbd>   | launch unit tests
<kbd>C-c , g</kbd>   | launch unit tests with tag (find, regexp, ...)

* Tips:

If you want to launch a single unit test, add a specify tag :

```lisp
(ert-deftest test-foobar ()
  :tags '(current)
  ```

And launch it using : <kbd>C-c , g</kbd> and specify tag : *current*


## Support / Contribute

See [here](CONTRIBUTING.md)


## Changelog

A changelog is available [here](ChangeLog.md).


## License

See [LICENSE](LICENSE).


## Contact

Nicolas Lamirault <nicolas.lamirault@gmail.com>



[ripgrep.el]: https://github.com/nlamirault/ripgrep.el
[badge-license]: https://img.shields.io/badge/license-GPL_2-green.svg?style=flat
[LICENSE]: https://github.com/nlamirault/ripgrep.el/blob/master/LICENSE

[GNU Emacs]: https://www.gnu.org/software/emacs/
[MELPA]: https://melpa.org/
[Cask]: http://cask.github.io/
[Issue tracker]: https://github.com/nlamirault/ripgrep.el/issues

[overseer]: https://github.com/tonini/overseer.el

[ripgrep]: https://github.com/BurntSushi/ripgrep



ChangeLog.md
# ripgrep.el ChangeLog

# Version 0.4.0 (02/11/2017)

- [#19](https://github.com/nlamirault/ripgrep.el/issues/19): : add a prefix test-helper
- [#9](https://github.com/nlamirault/ripgrep.el/issues/9): Add *projectile-ripgrep* to Melpa
- [#10](https://github.com/nlamirault/ripgrep.el/issues/10): `Unknown flag: '--not-file-matches'

# Version 0.3.0 (11/16/2016)

- `#7`: invalid pattern in ripgrep-search-mode (thanks kanru)
- `#6`: Couldn't jump from search result (thanks kanru)
- Use --vimgrep as well as --no-heading (thanks jasonm23)

# Version 0.2.0 (09/28/2016)

- `#3`: Add projectile-ripgrep tool

# Version 0.1.0 (09/23/2016)

- Setup unit tests (use [ert-runner][] and [overseer][])
- Init project

[ert-runner]: https://github.com/rejeep/ert-runner.el
[overseer]: https://github.com/tonini/overseer.el


test/test-helper.el
;; test-helper.el --- Test helpers for ripgrep.el

;; Copyright (C) 2016, 2017 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;; Homepage: https://github.com/nlamirault/ripgrep.el

;;; License:

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ansi)
(require 'cl) ;; http://emacs.stackexchange.com/questions/2864/symbols-function-definition-is-void-cl-macroexpand-all-when-trying-to-instal
(require 'ert)
(require 'f)
(require 'undercover)

(setq debugger-batch-max-lines (+ 50 max-lisp-eval-depth)
      debug-on-error t)

(defvar username (getenv "HOME"))

(defconst ripgrep-testsuite-dir
  (f-parent (f-this-file))
  "The testsuite directory.")

(defconst ripgrep-source-dir
  (f-parent ripgrep-testsuite-dir)
  "The ripgrep.el source directory.")

(defconst ripgrep-sandbox-path
  (f-expand "sandbox" ripgrep-testsuite-dir)
  "The sandbox path for ripgrep.")

(defun cleanup-load-path ()
  "Remove home directory from 'load-path."
  (message (ansi-green "[ripgrep] Cleanup path"))
  (mapc #'(lambda (path)
            (when (string-match (s-concat username "/.emacs.d") path)
              (message (ansi-yellow "Suppression path %s" path))
              (setq load-path (delete path load-path))))
        load-path))

(defun load-unit-tests (path)
  "Load all unit test from PATH."
  (message (ansi-green "[ripgrep] Execute unit tests %s"
                       path))
  (dolist (test-file (or argv (directory-files path t "-test.el$")))
    (load test-file nil t)))


(defun load-library (file)
  "Load current library from FILE."
  (let ((path (s-concat ripgrep-source-dir file)))
    (message (ansi-yellow "[ripgrep] Load library from %s" path))
    (undercover "*.el" (:exclude "*-test.el"))
    (require 'ripgrep path)))


(defmacro with-test-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(unwind-protect
       (condition-case nil ;ex
           (let ((default-directory ripgrep-source-dir))
             ;; (unless (f-dir? ripgrep-sandbox-path)
             ;;   (f-mkdir ripgrep-sandbox-path))
             (cleanup-load-path)
             (load-library "/ripgrep.el")
             ,@body)
         )))


(provide 'test-helper)
;;; test-helper.el ends here


test/sift-version-test.el
;;; ripgrep-version-test.el --- Tests for version information

;; Copyright (C) Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;;; Code:


(require 'test-helper)

(ert-deftest ripgrep-library-version ()
  :tags '(version)
  :expected-result (if (executable-find "cask") :passed :failed)
  (let* ((cask-version (car (process-lines "cask" "version"))))
    (message "ripgrep.el Cask version: %s" cask-version)
    (should (string= "0.2.0" cask-version))))


(provide 'ripgrep-version-test)
;;; ripgrep-version-test.el ends here


test/ripgrep-test-helper.el
;; ripgrep-test-helper.el --- Test helpers for ripgrep.el

;; Copyright (C) 2016, 2017 Nicolas Lamirault <nicolas.lamirault@gmail.com>

;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;; Homepage: https://github.com/nlamirault/ripgrep.el

;;; License:

;; This file is NOT part of GNU Emacs.

;; This program is free software; you can redistribute it and/or modify
;; it under the terms of the GNU General Public License as published by
;; the Free Software Foundation, either version 3 of the License, or
;; (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program.  If not, see <http://www.gnu.org/licenses/>.

;;; Commentary:

;;; Code:

(require 'ansi)
(require 'cl) ;; http://emacs.stackexchange.com/questions/2864/symbols-function-definition-is-void-cl-macroexpand-all-when-trying-to-instal
(require 'ert)
(require 'f)
(require 'undercover)

(setq debugger-batch-max-lines (+ 50 max-lisp-eval-depth)
      debug-on-error t)

(defvar ripgrep-test-helper-username (getenv "HOME"))

(defconst ripgrep-testsuite-dir
  (f-parent (f-this-file))
  "The testsuite directory.")

(defconst ripgrep-source-dir
  (f-parent ripgrep-testsuite-dir)
  "The ripgrep.el source directory.")

(defconst ripgrep-sandbox-path
  (f-expand "sandbox" ripgrep-testsuite-dir)
  "The sandbox path for ripgrep.")

(defun ripgrep-test-helper-cleanup-load-path ()
  "Remove home directory from 'load-path."
  (message (ansi-green "[ripgrep] Cleanup path"))
  (mapc #'(lambda (path)
            (when (string-match (s-concat ripgrep-test-helper-username "/.emacs.d") path)
              (message (ansi-yellow "Suppression path %s" path))
              (setq load-path (delete path load-path))))
        load-path))

(defun ripgrep-test-helper-load-unit-tests (path)
  "Load all unit test from PATH."
  (message (ansi-green "[ripgrep] Execute unit tests %s"
                       path))
  (dolist (test-file (or argv (directory-files path t "-test.el$")))
    (load test-file nil t)))


(defun ripgrep-test-helper-load-library (file)
  "Load current library from FILE."
  (let ((path (s-concat ripgrep-source-dir file)))
    (message (ansi-yellow "[ripgrep] Load library from %s" path))
    (undercover "*.el" (:exclude "*-test.el"))
    (require 'ripgrep path)))


(defmacro ripgrep-test-helper-with-test-sandbox (&rest body)
  "Evaluate BODY in an empty sandbox directory."
  `(unwind-protect
       (condition-case nil ;ex
           (let ((default-directory ripgrep-source-dir))
             ;; (unless (f-dir? ripgrep-sandbox-path)
             ;;   (f-mkdir ripgrep-sandbox-path))
             (cleanup-load-path)
             (load-library "/ripgrep.el")
             ,@body)
         )))


(provide 'ripgrep-test-helper)
;;; ripgrep-test-helper.el ends here


projectile-ripgrep.el
;;; projectile-ripgrep.el --- Run ripgrep with Projectile

;; Copyright (C) 2016, 2017 Nicolas Lamirault <nicolas.lamirault@gmail.com>
;;
;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;; Version: 0.4.0
;; Keywords : ripgrep projectile
;; Homepage: https://github.com/nlamirault/ripgrep.el
;; Package-Requires: ((ripgrep "0.3.0") (projectile "0.14.0"))

;;; Commentary:

;; Please see README.md for documentation, or read it online at
;; https://github.com/nlamirault/ripgrep.el

;;; License:

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;; Usage :

;; M-x projectile-ripgrep

;;; Code:

(require 'projectile)
(require 'ripgrep)

;;;###autoload
(defun projectile-ripgrep (regexp)
  "Run a Ripgrep search with `REGEXP' rooted at the current projectile project root."
  (interactive
   (list
    (read-from-minibuffer "Ripgrep search for: " (thing-at-point 'symbol))))
  (ripgrep-regexp regexp
                  (projectile-project-root)
                  (mapcar (lambda (val) (concat "--glob !" val))
                          (append projectile-globally-ignored-files
                                  projectile-globally-ignored-directories))))


(provide 'projectile-ripgrep)
;;; projectile-ripgrep.el ends here


ripgrep.el
;;; ripgrep.el --- Front-end for ripgrep, a command line search tool

;; Copyright (C) 2016 Nicolas Lamirault <nicolas.lamirault@gmail.com>
;;
;; Author: Nicolas Lamirault <nicolas.lamirault@gmail.com>
;; Version: 0.5.0
;; Keywords : ripgrep ack pt ag sift grep search
;; Homepage: https://github.com/nlamirault/ripgrep.el

;;; Commentary:

;; Emacs front-end for ripgrep, a command line search tool

;; Installation:

;; ripgrep.el is available on the two major community maintained repositories
;; Melpa stable (https://stable.melpa.org), and Melpa (https://melpa.org)
;;
;; (add-to-list 'package-archives
;;              '("melpa" . "https://melpa.org/packages/") t)
;;
;; M-x package-install ripgrep

;;; License:

;; This program is free software; you can redistribute it and/or
;; modify it under the terms of the GNU General Public License
;; as published by the Free Software Foundation; either version 2
;; of the License, or (at your option) any later version.

;; This program is distributed in the hope that it will be useful,
;; but WITHOUT ANY WARRANTY; without even the implied warranty of
;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
;; GNU General Public License for more details.

;; You should have received a copy of the GNU General Public License
;; along with this program; if not, write to the Free Software
;; Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
;; 02110-1301, USA.

;;; Commentary:

;; Usage :

;; M-x ripgrep-regexp

;;; Code:

(require 'compile)
(require 'grep)
(require 'thingatpt)


;; Customization
;; --------------------------

(defgroup ripgrep nil
  "Ripgrep"
  :group 'tools
  :group 'matching)


(defcustom ripgrep-executable
  "rg"
  "Name of the ripgrep executable to use."
  :type 'string
  :group 'ripgrep)


(defcustom ripgrep-arguments
  (list "")
  "Default arguments passed to ripgrep."
  :type '(repeat (string))
  :group 'ripgrep)


(defcustom ripgrep-highlight-search t
  "Non-nil means we highlight the current search term in results.
This requires the ripgrep command to support --color-match, which is only in v0.14+"
  :type 'boolean
  :group 'ripgrep)


;; Faces
;; --------------------------


(defface ripgrep-hit-face '((t :inherit compilation-info))
  "Face name to use for ripgrep matches."
  :group 'ripgrep)


(defface ripgrep-match-face '((t :inherit match))
  "Face name to use for ripgrep matches."
  :group 'ripgrep)



;; Mode
;; --------------------------


(defvar ripgrep-search-finished-hook nil
  "Hook run when ripgrep completes a search in a buffer.")

(defun ripgrep/run-finished-hook (buffer how-finished)
  "Run the ripgrep hook to signal that the search has completed."
  (with-current-buffer buffer
    (run-hooks 'ripgrep-search-finished-hook)))

(defvar ripgrep-search-mode-map
  (let ((map (make-sparse-keymap)))
    (set-keymap-parent map compilation-minor-mode-map)
    (define-key map "p" 'compilation-previous-error)
    (define-key map "n" 'compilation-next-error)
    (define-key map "{" 'compilation-previous-file)
    (define-key map "}" 'compilation-next-file)
    (define-key map "k" '(lambda ()
                           (interactive)
                           (let ((kill-buffer-query-functions))
                             (kill-buffer))))
    map)
  "Keymap for ripgrep-search buffers.
`compilation-minor-mode-map' is a cdr of this.")


(define-compilation-mode ripgrep-search-mode "Ripgrep"
  "Ripgrep results compilation mode"
  (set (make-local-variable 'truncate-lines) t)
  (set (make-local-variable 'compilation-disable-input) t)
  (set (make-local-variable 'tool-bar-map) grep-mode-tool-bar-map)
  (let ((symbol 'compilation-ripgrep)
        (pattern '("^\\([^:\n]+?\\):\\([0-9]+\\):\\([0-9]+\\):" 1 2 3)))
    (set (make-local-variable 'compilation-error-regexp-alist) (list symbol))
    (set (make-local-variable 'compilation-error-regexp-alist-alist) (list (cons symbol pattern))))
  (set (make-local-variable 'compilation-error-face) 'ripgrep-hit-face)
  (add-hook 'compilation-filter-hook 'ripgrep-filter nil t))


;; Taken from grep-filter, just changed the color regex.
(defun ripgrep-filter ()
  "Handle match highlighting escape sequences inserted by the rg process.
This function is called from `compilation-filter-hook'."
  (when ripgrep-highlight-search
    (save-excursion
      (forward-line 0)
      (let ((end (point)) beg)
        (goto-char compilation-filter-start)
        (forward-line 0)
        (setq beg (point))
        ;; Only operate on whole lines so we don't get caught with part of an
        ;; escape sequence in one chunk and the rest in another.
        (when (< (point) end)
          (setq end (copy-marker end))
          ;; Highlight rg matches and delete marking sequences.
          (while (re-search-forward "\033\\[30;43m\\(.*?\\)\033\\[[0-9]*m" end 1)
            (replace-match (propertize (match-string 1)
                                       'face nil 'font-lock-face 'ripgrep-match-face)
                           t t))
          ;; Delete all remaining escape sequences
          (goto-char beg)
          (while (re-search-forward "\033\\[[0-9;]*[mK]" end 1)
            (replace-match "" t t)))))))



;; API
;; --------------------------


;;;###autoload
(defun ripgrep-regexp (regexp directory &optional args)
  "Run a ripgrep search with `REGEXP' rooted at `DIRECTORY'.
`ARGS' provides Ripgrep command line arguments."
  (interactive
   (list (read-from-minibuffer "Ripgrep search for: " (thing-at-point 'symbol))
         (read-directory-name "Directory: ")))
  (let ((default-directory directory))
    (compilation-start
     (mapconcat 'identity
                (append (list ripgrep-executable)
                        ripgrep-arguments
                        args
                        '("--no-heading --vimgrep")
                        (list (shell-quote-argument regexp) ".")) " ")
     'ripgrep-search-mode)))


(provide 'ripgrep)
;;; ripgrep.el ends here
